### ISSUE TITLE
Ember 2.10.0 helper recompute bug

### ISSUE THREAD

**Issue by ssured** [emberjs/ember.js#14774]
Calling recompute on a helper does not recompute in 2.10.0

Twiddle: https://ember-twiddle.com/3afc1bda1c824f4ed141b666e5521320?openFiles=helpers.my-helper.js%2C

When you change the dependency to 2.9 this code works, for 2.10 it does not work. It might be related to Glimmer update?

**Comment by rwjblue** [comment]
Definitely seems funky, I agree.  The work around for now is to ensure the recompute is done within a run-loop (which fixes the twiddle):

```js
export default Ember.Helper.extend({
  count: 0,
  compute() {
    if (!this.timer) {
      this.timer = window.setInterval(() => {
        Ember.run.join(() => {
          this.incrementProperty('count');
          console.log('call recompute', this.get('count'));
      	  this.recompute();
        });
      }, 500);
    }    
    return "count = " + this.get('count');
  }
});
```

Off the top of my head, I am not sure why a run-loop is required for it to function, but wrapping sets/rerenders in a `run.join` like this is almost certainly better for testing (would prevent an auto-run assertion in tests) so it is a reasonable solution.

**Comment by youroff** [comment]
Why is this closed? Seems like an actual bug or what's the rationale behind the need of manually wrapping `recompute` call in `run.join`?

**Comment by pixelhandler** [comment]
@locks @rwjblue @ssured @youroff is this still an issue, perhaps we should close or create a new reproduction of this, what do you think?

**Comment by boris-petrov** [comment]
@pixelhandler - we just hit the same issue in Ember 3.4.5. So it is still valid. We had to wrap the call to `recompute` in a `run`. I created [another issue a few days ago](https://github.com/emberjs/ember.js/issues/17021) which is very similar to this problem that we face here. I.e. the code we had was:

```coffeescript
export default Helper.extend
  compute: ([a, b]) ->
    pendingResult = @get('_pendingResult')
    if pendingResult?
      @set('_pendingResult', undefined)
      return pendingResult

    result = a.someMethodWithNativeAwaitsInside(b)

    result.then (result) =>
      @set('_pendingResult', result)
      @recompute()

    false
```

And this "blocked" when `recompute` was called - i.e. it didn't call `compute` again. And this started appearing only after we changed our code to use native promises instead of RSVP promises (or most of it - because of the other issue I mentioned). So there is some inconsistency between RSVP promises and native promises (under the latest Chrome) - and because of it, stuff breaks.

Any ideas?

**Comment by rwjblue** [comment]
The solution here is to ensure we `run.join` inside the `recompute` code (as @ssured 	mentioned in a prior comment).

**Comment by rwjblue** [comment]
I don't understand what you mean @boris-petrov. I'm saying that Ember.Helper's implementation of `recompute` should be using `run.join` not that your user-land helper code should...

**Comment by boris-petrov** [comment]
Ah, sorry, I thought you meant putting “run” around the “recompute” call in the user code as your initial suggestion. I misread, sorry!

But the other issue still remains.

**Comment by xg-wang** [comment]
This issue can be worked around by creating a new runloop as explained in https://github.com/rwjblue/ember-router-helpers/pull/126.
@rwjblue should `recompute`'s internal impl be using runloop or just let userland code doing the `run.join` workaround?

**Comment by buschtoens** [comment]
I also just ran into this, because `ember-on-modifier` does not create a new runloop. I'm not really sure, whether we should consider this a bug in the framework or expected behavior.

**Comment by rwjblue** [comment]
Definitely a bug in the framework. We need to fix 😢.

**Comment by buschtoens** [comment]
> I'm saying that Ember.Helper's implementation of recompute should be using run.join not that your user-land helper code should...

Easy PR right there! I'll take it tomorrow, if nobody else signs up. 😏
